### PR TITLE
Fix shellcheck warnings/errors in cluster/gce/util.sh

### PR DESCRIPTION
**Pull request by joakimr-axis**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Fixes shellcheck failures in cluster/gce/util.sh

**Which issue(s) this PR fixes**:
ref: #72956

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: cluster/gce/util.sh
Comment (spiffxp) [inline cluster/gce/util.sh:3519]: are you sure we need this? I suspect we may want globbing

File: cluster/gce/util.sh
Comment (spiffxp) [inline cluster/gce/util.sh:3765]: this has the potential to change the exit code, are we sure we want to do that?

**Review by spiffxp — COMMENTED** [review]
/hold
Sorry, but I am not willing to accept a size/XXL change against a bash script in cluster without very good reasons.  Is it really necessary to change all the `"` marks to `'`?

File: cluster/gce/util.sh
Comment (joakimr-axis) [reply cluster/gce/util.sh:3765]: Ah, good point. It does seem strange that a function should pretend to successfully return something correctly when it is failing, but the purpose of this patch was not to change the behavior. I will change this back to the echo and add a comment + a shellcheck disable.

**Comment by joakimr-axis** [comment]
> /hold
> Sorry, but I am not willing to accept a size/XXL change against a bash script in cluster without very good reasons. Is it really necessary to change all the `"` marks to `'`?

No, for sure it is not necessary just to pass hack/verify-shellcheck. With the massive amount of quote suggestions from shellcheck this was the result of going through them and sorting out what is/was what. I will create and update where this is purged.

File: cluster/gce/util.sh
Comment (joakimr-axis) [reply cluster/gce/util.sh:3519]: For array entries where there is no whitespace, the outcome will be the same. But for entries where there is whitespace (although I guess there are no such groups in reality?), there is indeed a difference: I am under the impression that each array entry contains 1 group name (in which case globbing is not wanted), where a group name if it contains a whitespace with the current code on master will be split.

```
INSTANCE_GROUPS=('group 0')
INSTANCE_GROUPS+=('group 1')
WINDOWS_INSTANCE_GROUPS=('wingroup a')
WINDOWS_INSTANCE_GROUPS+=('wingroup b')

# Current code on master
all_instance_groups=(${INSTANCE_GROUPS[@]:-} ${WINDOWS_INSTANCE_GROUPS[@]:-})
for group in ${all_instance_groups[@]:-}; do
        echo "$group"
done

echo '**********'

# Updated code in my patch
new_all_instance_groups=("${INSTANCE_GROUPS[@]:-}" "${WINDOWS_INSTANCE_GROUPS[@]:-}")
for group in "${new_all_instance_groups[@]:-}"; do
        echo "$group"
done
```

The output would be

```
group
0
group
1
wingroup
a
wingroup
b
**********
group 0
group 1
wingroup a
wingroup b
```

**Comment by joakimr-axis** [comment]
> Sorry, but I am not willing to accept a size/XXL change against a bash script in cluster without very good reasons.?

The current updated patchset contains nothing more than the lines needed to be addressed in order for the file to pass hack/verify-shellcheck. It is still size/XXL, I am afraid. Please let me know if you would like to split it into sub-PRs (e.g. "fix unnecessary cmd" in one, "fix quotations" in one, "fix mapfile/read" in one, etc.).

**Review by spiffxp — CHANGES_REQUESTED** [review]
I finally took a pass at this.  This is not easy to review and honestly I'm still not sure it's worth the risk.  I can take another look if you address comments, but I'm not guaranteeing this will merge

File: cluster/gce/util.sh
Comment (joakimr-axis) [reply cluster/gce/util.sh:2570]: I guess since the firewall rules will probably never contain things like whitespaces, we will never see a bug here. But if they did, the current code would split such an argument.

File: cluster/gce/util.sh
Comment (joakimr-axis) [reply cluster/gce/util.sh:3535]: Sure thing. Will add shellcheck disable comment instead.

File: cluster/gce/util.sh
Comment (joakimr-axis) [reply cluster/gce/util.sh:1620]: You are right. There is a period missing. Thanks for spotting.

File: cluster/gce/util.sh
Comment (joakimr-axis) [reply cluster/gce/util.sh:2013]: Fix upcoming.

File: cluster/gce/util.sh
Comment (joakimr-axis) [reply cluster/gce/util.sh:414]: I interpreted this usecase as the case described as

"Alternatively, if you do want to concatenate all the values in the array into a single string for your test, use `"$*"` or `"${array[*]}"` to make this explicit."

in https://github.com/koalaman/shellcheck/wiki/SC2199.

File: cluster/gce/util.sh
Comment (joakimr-axis) [reply cluster/gce/util.sh:402]: I interpreted this usecase as the case described as

"Alternatively, if you do want to concatenate all the values in the array into a single string for your test, use "$*" or "${array[*]}" to make this explicit."

in https://github.com/koalaman/shellcheck/wiki/SC2199.

File: cluster/gce/util.sh
Comment (joakimr-axis) [reply cluster/gce/util.sh:2477]: Should not be there. Thanks for spotting.

File: cluster/gce/util.sh
Comment (joakimr-axis) [reply cluster/gce/util.sh:416]: The current way of adding the elements from the gcloud command to the array WINDOWS_NODE_NAMES trigges https://github.com/koalaman/shellcheck/wiki/SC2207. Using mapfile for the assignment of one or multiple elements from the gcloud comand is a suggested alternative to mitigate that. The `-O` option to mapfile is for appending to existing content in the variable (https://www.computerhope.com/unix/bash/mapfile.htm).

File: cluster/gce/util.sh
Comment (joakimr-axis) [reply cluster/gce/util.sh:404]: The current way of adding the elements from the gcloud command to the array `NODE_NAMES` trigges https://github.com/koalaman/shellcheck/wiki/SC2207. Using `mapfile` for the assignment of one or multiple elements from the gcloud comand is a suggested alternative to mitigate that. The `-O` option to mapfile is for appending to existing content in the variable (https://www.computerhope.com/unix/bash/mapfile.htm).

**Comment by joakimr-axis** [comment]
> I finally took a pass at this. This is not easy to review and honestly I'm still not sure it's worth the risk. I can take another look if you address comments, but I'm not guaranteeing this will merge

This is not my baby; I am only doing this trying to help the guys working with https://github.com/kubernetes/kubernetes/issues/72956. Since you (as the k8s guys) have decided to have shellcheck for your shell scripts and to create that PR, I believe the aim is to get all scripts in line?
As I have suggested in a previous comment here, splitting the fixes for cluster/gce/util.sh into several patches might be a more convenient way forward to that, but have received no response on that suggestion so far.

File: cluster/gce/util.sh
Comment (joakimr-axis) [reply cluster/gce/util.sh:404]: An alternative to mapfile could be using read instead, making it friendlier to legacy bash versions. Something like

`while IFS= read -r line; do NODE_NAMES+=("$line"); done <<< "$(gcloud compute ...`

File: cluster/gce/util.sh
Comment (joakimr-axis) [reply cluster/gce/util.sh]: Replaced mapfile with read

**Comment by joakimr-axis** [comment]
I have now split parts from this currently huge PR into a number of smaller ones, as requested by @BenTheElder in https://github.com/kubernetes/kubernetes/issues/72956#issuecomment-686078249.

Please see:
#94456 (shellcheck disables where we deliberately want to have the current code that shellcheck triggers on)
#94457 (remove superfluous use of cat)
#94458 (clean up the use of arithmetic shell functionality)
#94459 (split declaration and assignment for local variables, add missing exports, remove a unused variable)

**Comment by cheftako** [comment]
/assign @spiffxp 
While I appreciate the fine work you did here, I have to agree with Aaron.
Can we break this PR into smaller PRs which focus on individual classes of error.
If we do find a problem it would allow us to roll back just a portion of the changes.
I do realize we cannot enable shellcheck until all the pieces land.

**Comment by joakimr-axis** [comment]
> /assign @spiffxp
> While I appreciate the fine work you did here, I have to agree with Aaron.
> Can we break this PR into smaller PRs which focus on individual classes of error.

But it _is_ broken down into smaller PRs already!
Please see my comment above from early September (https://github.com/kubernetes/kubernetes/pull/90405#issuecomment-686323844).

Several of these sub-PRs have already been merged since, with

#94463
#94459
#94456

remaining.

With those 3 merged, the leftover fixes would be found here and presumably few enough to make it all mangeable.

**Comment by saschagrunert** [comment]
I removed that from the release engineering project board because I think the modified code parts do not need approval from us. Feel free to chime in if this is not the case.

**Comment by joakimr-axis** [comment]
> Several of these sub-PRs have already been merged since, with
> 
> #94463
> #94459
> #94456
> 
> remaining.

Currently, only #94459 remaining.

**Comment by joakimr-axis** [comment]
> Currently, only #94459 remaining.

Now all those sub-PRs have been merged. What do you think, should I split this one into more parts (I can do that) or should we fly with as it is now (size/L)?

Ping @BenTheElder.

**Comment by BenTheElder** [comment]
this diff is much smaller now, I'll start reviewing it again.

File: cluster/gce/util.sh
Comment (BenTheElder) [reply cluster/gce/util.sh:2383]: heh, the `|| true` here is here because we store the value and then check if empty below (and then create the subnet if so), we don't want to fail if the subnet doesn't exist yet.

**Comment by BenTheElder** [comment]
/retest
https://github.com/kubernetes/kubernetes/issues/98893

**Comment by BenTheElder** [comment]
... neat. Just found https://github.com/kubernetes/kubernetes/issues/98941

/retest

**Pull request merged**
